lib/http_v4_signature.c

```diff
@@ -0,0 +1,329 @@
+  char up_provider[PROVIDER_MAX_L + 1] = {0};
+  char mid_provider[PROVIDER_MAX_L + 1] = {0};
+  char *region = NULL;
+  char *uri = NULL;
+  char *api_type = NULL;
+  char date_iso[17];
+  char date[9];
+  char date_str[64];
+  const char *post_data = data->set.postfields ?
+    data->set.postfields : "";
+  const char *content_type = Curl_checkheaders(conn, "Content-Type");
+  unsigned char sha_d[32];
+  char sha_str[65];
+  char *cred_scope = NULL;
+  char *signed_headers = NULL;
+  char request_type[REQUEST_TYPE_L];
+  char *canonical_hdr = NULL;
+  char *canonical_request = NULL;
+  char *str_to_sign = NULL;
+  unsigned char tmp_sign0[32] = {0};
+  unsigned char tmp_sign1[32] = {0};
+  char *auth = NULL;
+  char *tmp;
+  #ifdef DEBUGBUILD
+  char *force_timestamp;
+  #endif
+
+  DEBUGASSERT(!proxy);
+  (void)proxy;
+
+  if(Curl_checkheaders(conn, "Authorization")) {
+    /* Authorization already present, Bailing out */
+    return CURLE_OK;
+  }
+
+  if(!customrequest)
+    customrequest = "POST";
+  if(content_type) {
+    content_type = strchr(content_type, ':');
+    if(!content_type)
+      return CURLE_FAILED_INIT;
+    content_type++;
+    /* Skip whitespace now */
+    while(*content_type == ' ' || *content_type == '\t')
+      ++content_type;
+  }
+
+  /* Get Parameter
+     Google and Outscale use the same OSC or GOOG,
+     but Amazon use AWS and AMZ for header arguments */
+  tmp = strchr(provider, ':');
+  if(tmp) {
+    provider_l = tmp - provider;
+    if(provider_l >= PROVIDER_MAX_L) {
+      infof(data, "v4 signature argument string too long\n");
+      return CURLE_BAD_FUNCTION_ARGUMENT;
+    }
+    Curl_strntolower(low_provider0, provider, provider_l);
+    Curl_strntoupper(up_provider, provider, provider_l);
+    provider = tmp + 1;
+    /* if "xxx:" was pass as parameter, tmp + 1 should point to \0 */
+    provider_l = strlen(provider);
+    if(provider_l >= PROVIDER_MAX_L) {
+      infof(data, "v4 signature argument string too long\n");
+      return CURLE_BAD_FUNCTION_ARGUMENT;
+    }
+    Curl_strntolower(low_provider, provider, provider_l);
+    Curl_strntolower(mid_provider, provider, provider_l);
+  }
+  else if(provider_l <= PROVIDER_MAX_L) {
+    Curl_strntolower(low_provider0, provider, provider_l);
+    Curl_strntolower(low_provider, provider, provider_l);
+    Curl_strntolower(mid_provider, provider, provider_l);
+    Curl_strntoupper(up_provider, provider, provider_l);
+    mid_provider[0] = Curl_raw_toupper(mid_provider[0]);
+  }
+  else {
+    infof(data, "v4 signature argument string too long\n");
+    return CURLE_BAD_FUNCTION_ARGUMENT;
+  }
+
+#ifdef DEBUGBUILD
+  force_timestamp = getenv("CURL_FORCETIME");
+  if(force_timestamp)
+    rawtime = 0;
+  else
+#endif
+    time(&rawtime);
+
+  ret = Curl_gmtime(rawtime, &info);
+  if(ret != CURLE_OK) {
+    return ret;
+  }
+
+  if(!strftime(date_iso, 17, "%Y%m%dT%H%M%SZ", &info)) {
+    return CURLE_OUT_OF_MEMORY;
+  }
+  date_iso[16] = 0;
```

Comment:
Here too, can we have `16` be sizeof -1? Hardcoded numbers are asking for future problems,